### PR TITLE
More Accumulated Data

### PR DESCRIPTION
**Pull request by lkiesow**
This patch makes the API return more accumulated data (min, max,
standard deviation). It also delivers the data in distinct categories.

Finally, this moves the dict generation into the database code since it
is highly specific to the result of the database query.

### CODE REVIEW

**Comment by shaardie** [comment]
It is a nice idea to add more aggregated data. I have not looked into in detail, but I have some comments about this.

All functions, the endpoint and the documentation are about the averages, but now there is much more about it. I think this is misleading. If we want to do this, we should change those names and descriptions.

Also I am wondering how much slower it is to get all there aggregated values in comparison with only the average. If this is significant more? We should think about the losses and benefits here.

And last our current functions in `emissionsapi/db.py` are returning the queries and not the results. Therefore different endpoint could use those functions differently and chain them at will. We do not have to stick to this schema, but if we change this we should have a reason for that.

**Comment by lkiesow** [comment]
I went back to returning a query to make you happy though I really dislike returning this magic data about which the receiver needs very specific, internal knowledge.

I've also renamed the methods to *daily* instead of average though I would have seen the other aggregated data having a relation to this and wouldn't have minded keeping that name.

The documentation is updated.

> Also I am wondering how much slower it is to get all the aggregated values in comparison to only the average. If this is significant more, we should think about the losses and benefits here.

There is basically no difference:

```sql
emissionsapi=> select avg(value), min(timestamp), max(timestamp), date_trunc('hour', timestamp) as day from carbonmonoxide group by date_trunc('hour', timestamp) order by day limit 10;
…
Time: 2064.946 ms (00:02.065)fg


emissionsapi=> select count(value), avg(value), min(timestamp), max(timestamp), min(value), max(value), stddev(value), date_trunc('hour', timestamp) as day from carbonmonoxide group by date_trunc('hour', timestamp) order by day limit 10;
…
Time: 2035.302 ms (00:02.035)
```

File: emissionsapi/openapi.yml [hunk 1]
```diff
@@ -35,18 +35,18 @@ paths:
...
             application/json:
               schema:
                 $ref: '#/components/schemas/Feature Collection'
         400:
           description: Parameter error
-  /api/v1/average.json/:
+  /api/v1/daily.json:
     get:
-      operationId: emissionsapi.web.get_average
+      operationId: emissionsapi.web.get_daily
```
Comment (shaardie) [inline emissionsapi/openapi.yml]:
```suggestion
      operationId: emissionsapi.web.get_daily_statistics
```

File: emissionsapi/openapi.yml [hunk 2]
```diff
@@ -35,18 +35,18 @@ paths:
...
       description: |
-          Get daily average for a specified area filtered by time.
+          Get daily values for a specified area filtered by time.
```
Comment (shaardie) [inline emissionsapi/openapi.yml]:
```suggestion
          Get daily statistics for a specified area filtered by time.
```

File: emissionsapi/openapi.yml [hunk 3]
```diff
@@ -63,11 +63,11 @@ paths:
         - $ref: '#/components/parameters/offset'
       responses:
         200:
-          description: Array of calculated averages.
+          description: Array of daily carbon monoxide values.
```
Comment (shaardie) [inline emissionsapi/openapi.yml]:
```suggestion
          description: Array of daily carbon monoxide statistics.
```

File: emissionsapi/openapi.yml [hunk 4]
```diff
@@ -63,11 +63,11 @@ paths:
...
           content:
             application/json:
               schema:
-                $ref: '#/components/schemas/Averages'
+                $ref: '#/components/schemas/Daily'
```
Comment (shaardie) [inline emissionsapi/openapi.yml]:
```suggestion
                $ref: '#/components/schemas/Daily Statistics'
```

File: emissionsapi/openapi.yml [hunk 5]
```diff
@@ -174,18 +174,39 @@ components:
                   type:  # This is the name of the key
                     type: string
                     example: "Point"
-    Averages:
+    Daily:
```
Comment (shaardie) [inline emissionsapi/openapi.yml]:
```suggestion
    Daily Statistics:
```

File: emissionsapi/openapi.yml [hunk 6]
```diff
@@ -174,18 +174,39 @@ components:
...
       type: array
-      description: Array of average values.
+      description: Array of daily carbon monoxide values.
```
Comment (shaardie) [inline emissionsapi/openapi.yml]:
```suggestion
      description: Array of daily carbon monoxide statistics.
```

File: emissionsapi/openapi.yml [hunk 7]
```diff
@@ -174,18 +174,39 @@ components:
...
-          average:
-            type: number
-            example: 0.0312481193586528
-          start:
-            type: date-time
-            example: 2019-02-10T13:05:08.812000Z
-          end:
-            type: date-time
-            example: 2019-02-10T13:03:40.254000Z
+          value:
```
Comment (shaardie) [inline emissionsapi/openapi.yml]:
```suggestion
          statistics:
```

File: emissionsapi/web.py [hunk 8]
```diff
@@ -106,8 +106,8 @@ def get_data(session, country=None, geoframe=None, polygon=None,
 
 @parse_date('begin', 'end')
 @emissionsapi.db.with_session
-def get_average(session, country=None, geoframe=None, polygon=None,
-                begin=None, end=None, limit=None, offset=None):
+def get_daily(session, country=None, geoframe=None, polygon=None,
```
Comment (shaardie) [inline emissionsapi/web.py]:
```suggestion
def get_daily_statistics(session, country=None, geoframe=None, polygon=None,
```

Comment (shaardie) [reply emissionsapi/web.py]: The following line has to be align correctly again...

File: emissionsapi/web.py [hunk 9]
```diff
@@ -127,19 +127,23 @@ def get_average(session, country=None, geoframe=None, polygon=None,
         except RESTParamError as err:
             return str(err), 400
 
-    query = emissionsapi.db.get_averages(
+    query = emissionsapi.db.get_daily(
```
Comment (shaardie) [inline emissionsapi/web.py]:
```suggestion
    query = emissionsapi.db.get_daily_statistics(
```

File: emissionsapi/web.py [hunk 10]
```diff
@@ -127,19 +127,23 @@ def get_average(session, country=None, geoframe=None, polygon=None,
...
         query, limit=limit, offset=offset)
-
-    result = []
-    for avg, max_time, min_time, _ in query:
-        result.append({
-            'average': avg,
-            'start': min_time,
-            'end': max_time})
-    return result
+    return [{'value': {
```
Comment (shaardie) [inline emissionsapi/web.py]:
```suggestion
    return [{'statistics': {
```

**Comment by tibroc** [comment]
I would like to make a case for calling the API endpoint 'daily_statistics' or something to that effect that includes the word 'statistics'. I find it more explicit and more descriptive than just 'values'. I also find it helpful to include the word 'daily'. It might just be a side effect that this is now a 'daily statistic', but it makes it a clear design decision that is communicated through the endpoint's name. 
I think we could also decide that this is our fines data resolution that we provide. It would be a reasonable simplification of the interface IMHO.

File: emissionsapi/db.py [hunk 11]
```diff
@@ -148,8 +148,9 @@ def get_points(session, polygon=None, begin=None, end=None):
     return filter_query(query, polygon=polygon, begin=begin, end=end)
 
 
-def get_averages(session, polygon=None, begin=None, end=None):
-    """Get daily averages of all points filtered by time and location.
+def get_daily(session, polygon=None, begin=None, end=None):
```
Comment (shaardie) [inline emissionsapi/db.py]:
```suggestion
def get_daily_statistics(session, polygon=None, begin=None, end=None):
```

File: emissionsapi/openapi.yml [hunk 12]
```diff
@@ -35,18 +35,18 @@ paths:
...
-          description: The Response contains all known points located within the specified rectangle contained in a GeoJSON feature collection.
+          description: The response contains all known points located within the specified rectangle contained in a GeoJSON feature collection.
           content:
             application/json:
               schema:
                 $ref: '#/components/schemas/Feature Collection'
         400:
           description: Parameter error
-  /api/v1/average.json/:
+  /api/v1/daily.json:
```
Comment (shaardie) [inline emissionsapi/openapi.yml]:
```suggestion
  /api/v1/daily_statistics.json:
```

[the remark this answers is not in the document]
Comment (lkiesow) [reply emissionsapi/openapi.yml]: This has been addressed separately in pull request #127

**Pull request closed without merging**